### DISCUSSION TITLE
SSR: redirect(), notfound() not working of next/navigation

### DISCUSSION THREAD

**Discussion by ekaterina4952** [vercel/next.js#49835] [Help] [answered]
### Summary

I am trying to redirect directly on the page (ssr) such as middleware.(nextjs 13.4.1)

in page
```
const Index = ({businessInfo}) => {
    if (!businessInfo) redirect('/page')
    return <></>
}
```
I was curious it may works in `getServerSideProps`
```
export async function getServerSideProps(context) {
    if (!context.query.name) {
      redirect('/page')
    }
}
```
But I am failure both case. It also happens even though I use `notfound()`.

```
**Server Error**
**Error: NEXT_REDIRECT**
This error happened while generating the page. Any console logs will be displayed in the terminal window.
```

**Comment by icyJoseph** [comment, marked as answer]
Hi

I think you are mixing things up... those functions from `next/navigate` only work in the `app` Dir technology stack, while it seems that you are using the `pages` Dir technology stack.

Here you can find how to do redirect and notfound in the `pages` dir tech stack: https://nextjs.org/docs/pages/api-reference/functions/get-server-side-props#getserversideprops-return-values

For example:

```js
export async function getServerSideProps(context) {
  const res = await fetch(`https://.../data`);
  const data = await res.json();
 
  if (!data) {
    return {
      notFound: true,
    };
  }
 
  return {
    props: { data }, // will be passed to the page component as props
  };
}
```

If you want to redirect within a component, I'd first recommend to do it in GetServerSideProps, but if you really want to, you do that inside a `use-effect` using `useRouter()` - example:

```jsx
import { useEffect } from 'react';
import { useRouter } from 'next/router';
 
// Here you would fetch and return the user
const useUser = () => ({ user: null, loading: false });
 
export default function Page() {
  const { user, loading } = useUser();
  const router = useRouter();
 
  useEffect(() => {
    async function verifyUser() {
      if (!(user || loading)) {
        await router.push('/login');
      }
    }
    void verifyUser();
  }, [user, loading]);
 
  return <p>Redirecting...</p>;
}
```

**Reply by ekaterina4952** [reply]
you're the best!
